cpp/src/lists/extract.cu

```diff
@@ -32,36 +32,70 @@ namespace {
 /**
  * @brief Convert index value for each sublist into a gather index for
  * the lists column's child column.
+ * This specialization is for when the index is a fixed size_type value.
  */
-template <bool PositiveIndex = true>
+template <bool PositiveIndex = true, typename IndexType = size_type>
 struct map_index_fn {
   column_device_view const d_offsets;  // offsets to each sublist (including validity mask)
-  size_type const index;               // index of element within each sublist
+  IndexType const index;               // index of element within each sublist
   size_type const out_of_bounds;       // value to use to indicate out-of-bounds
 
   __device__ int32_t operator()(size_type idx)
   {
     if (d_offsets.is_null(idx)) return out_of_bounds;
     auto const offset = d_offsets.element<int32_t>(idx);
     auto const length = d_offsets.element<int32_t>(idx + 1) - offset;
-    if (PositiveIndex)
+    if constexpr (PositiveIndex) { return index < length ? index + offset : out_of_bounds; }
+    return index >= -length ? length + index + offset : out_of_bounds;
```

Comment:
I believe this is far more complicated than it needs to be. This looks like precisely the same logic as what is done in [`gather`](https://github.com/rapidsai/cudf/blob/4349232eb68abd3255bf677cee5456a7d4684e20/cpp/src/copying/gather.cu#L46) 
```suggestion
   auto const wrapped = (index < = 0) ? index + length + offset : index + offset;
   return (wrapped >= 0 and wrapped < length) ? wrapped : out_of_bounds; 
```

Reply:
The `PositiveIndex`, etc. is the code I'm trying to retain for `extract_list_element(list_column_view, size_type, mr)`. I interpreted your advice [here](https://github.com/rapidsai/cudf/issues/9172#issuecomment-911902656) to be to retain the algo and fit the `column_view` version around it.